core/src/main/scala/org/apache/spark/crypto/CryptoInputStream.scala

```diff
@@ -0,0 +1,439 @@
+
+  /**
+   * This method is executed immediately after decryption. Check whether
+   * decryptor should be updated and recalculate padding if needed.
+   */
+  def afterDecryption(decryptor: Decryptor, inBuffer: ByteBuffer, position: Long,
+                      iv: Array[Byte]): Byte = {
+    var padding: Byte = 0
+    if (decryptor.isContextReset) {
+      updateDecryptor(decryptor, position, iv)
+      padding = getPadding(position)
+      inBuffer.position(padding)
+    }
+    padding
+  }
+
+  def getCounter(position: Long): Long = {
+    position / codec.getCipherSuite.algoBlockSize
+  }
+
+  def getPadding(position: Long): Byte = {
+    (position % codec.getCipherSuite.algoBlockSize).asInstanceOf[Byte]
+  }
+
+  /** Calculate the counter and iv, update the decryptor. */
+  def updateDecryptor(decryptor: Decryptor, position: Long, iv: Array[Byte]) {
+    val counter: Long = getCounter(position)
+    codec.calculateIV(initIV, counter, iv)
+    decryptor.init(key, iv)
+  }
+
+  /**
+   * Reset the underlying stream offset clear {@link #inBuffer} and
+   * {@link #outBuffer}. This Typically happens during {@link #seek(long)}
+   * or {@link #skip(long)}.
+   */
+  def resetStreamOffset(offset: Long) {
+    streamOffset = offset
+    inBuffer.clear
+    outBuffer.clear
+    outBuffer.limit(0)
+    updateDecryptor(decryptor, offset, iv)
+    padding = getPadding(offset)
+    inBuffer.position(padding)
+  }
+
+  override def close {
+    if (!closed) {
+      super.close
+      freeBuffers
+      closed = true
+    }
+  }
+
+  /** Skip n bytes */
+  override def skip(nVal: Long): Long = {
+    var n: Long = nVal
+    Preconditions.checkArgument(n >= 0, "Negative skip length.", new Array[String](1))
+    checkStream
+    if (n == 0) {
+      0
+    }
+    else if (n <= outBuffer.remaining) {
+      val pos: Int = outBuffer.position + n.asInstanceOf[Int]
+      outBuffer.position(pos)
+      n
+    }
+    else {
+      n -= outBuffer.remaining
+      var skipped: Long = in.skip(n)
+      if (skipped < 0) {
+        skipped = 0
+      }
+      val pos: Long = streamOffset + skipped
+      skipped += outBuffer.remaining
+      resetStreamOffset(pos)
+      skipped
+    }
+  }
+
+  /** ByteBuffer read. */
+  override def read(buf: ByteBuffer): Int = {
+    checkStream
+    if (isReadableByteChannel) {
+      val unread: Int = outBuffer.remaining
+      if (unread > 0) {
+        val toRead: Int = buf.remaining
+        if (toRead <= unread) {
+          val limit: Int = outBuffer.limit
+          outBuffer.limit(outBuffer.position + toRead)
+          buf.put(outBuffer)
+          outBuffer.limit(limit)
+          toRead
+        }
+        else {
+          buf.put(outBuffer)
+        }
+      }
+      val pos: Int = buf.position
+      val n: Int = (in.asInstanceOf[ReadableByteChannel]).read(buf)
+      if (n > 0) {
+        streamOffset += n
+        decrypt(buf, n, pos)
+      }
+      if (n >= 0) {
+        unread + n
+      }
+      else {
```

Comment:
nit: move to previous line. Also in lots of other places in the code. I'll just stop pointing these out.